pkg/scheduling/v1/lease_manager.go

```diff
@@ -325,78 +295,82 @@ func (l *LeaseManager) loopForLeases(ctx context.Context) {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
+			// we acquire a processMu lock here to prevent cleanup from occurring simultaneously
+			l.processMu.Lock()
+
+			// we don't want to block the cleanup process, so we use a separate context with a timeout
+			loopCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+
 			wg := sync.WaitGroup{}
 
 			wg.Add(3)
 
 			go func() {
 				defer wg.Done()
-				if err := l.acquireWorkerLeases(ctx); err != nil {
+
+				if err := l.acquireWorkerLeases(loopCtx); err != nil {
 					l.conf.l.Error().Err(err).Msg("error acquiring worker leases")
 				}
 			}()
 
 			go func() {
 				defer wg.Done()
-				if err := l.acquireQueueLeases(ctx); err != nil {
+
+				if err := l.acquireQueueLeases(loopCtx); err != nil {
 					l.conf.l.Error().Err(err).Msg("error acquiring queue leases")
 				}
 			}()
 
 			go func() {
 				defer wg.Done()
-				if err := l.acquireConcurrencyLeases(ctx); err != nil {
+
+				if err := l.acquireConcurrencyLeases(loopCtx); err != nil {
 					l.conf.l.Error().Err(err).Msg("error acquiring concurrency leases")
 				}
 			}()
 
 			wg.Wait()
+
+			cancel()
+			l.processMu.Unlock()
 		}
 	}
 }
 
 func (l *LeaseManager) cleanup(ctx context.Context) error {
-	l.cleanupMu.Lock()
-	defer l.cleanupMu.Unlock()
+	// we acquire a process locks here to prevent concurrent cleanup and lease acquisition
+	l.processMu.Lock()
+	defer l.processMu.Unlock()
 
 	if l.cleanedUp {
 		return nil
 	}
 
 	l.cleanedUp = true
 
-	// close channels
-	defer close(l.workersCh)
-	defer close(l.queuesCh)
-	defer close(l.concurrencyLeasesCh)
-
 	eg := errgroup.Group{}
 
 	eg.Go(func() error {
-		l.workerLeasesMu.Lock()
-		defer l.workerLeasesMu.Unlock()
-
 		return l.lr.ReleaseLeases(ctx, l.tenantId, l.workerLeases)
 	})
 
 	eg.Go(func() error {
-		l.queueLeasesMu.Lock()
-		defer l.queueLeasesMu.Unlock()
-
 		return l.lr.ReleaseLeases(ctx, l.tenantId, l.queueLeases)
 	})
 
 	eg.Go(func() error {
-		l.concurrencyLeasesMu.Lock()
-		defer l.concurrencyLeasesMu.Unlock()
-
 		return l.lr.ReleaseLeases(ctx, l.tenantId, l.concurrencyLeases)
 	})
 
 	if err := eg.Wait(); err != nil {
 		return err
 	}
 
+	// close channels: this is safe to do because each channel is guarded by l.cleanedUp + the process lock
+	close(l.workersCh)
```

Comment:
Removed the `defer` when closing the channels since it's more difficult to reason about Go's LIFO stacking of defers.